sources/packages/backend/apps/api/src/institution-integration/ier-integration/ier-request.service.ts

```diff
@@ -0,0 +1,155 @@
+import { Injectable } from "@nestjs/common";
+import { StudentAssessment } from "@sims/sims-db";
+import { InjectLogger } from "../../common";
+import { LoggerService } from "../../logger/logger.service";
+import { ConfigService, StudentAssessmentService } from "../../services";
+import { IERIntegrationConfig } from "../../types";
+import { getFileNameAsCurrentTimestamp } from "../../utilities";
+import { IERIntegrationService } from "./ier-integration.service";
+import { IERRecord, IERUploadResult } from "./models/ier-integration.model";
+
+@Injectable()
+export class IERRequestService {
+  ierIntegrationConfig: IERIntegrationConfig;
+  constructor(
+    config: ConfigService,
+    private readonly ierIntegrationService: IERIntegrationService,
+    private readonly studentAssessmentService: StudentAssessmentService,
+  ) {
+    this.ierIntegrationConfig = config.getConfig().IERIntegrationConfig;
+  }
+
+  /**
+   * 1. Fetches the assessment data for the institution location.
+   * 2. Create the Request content for the IER 12 file by populating the
+   * header, footer and trailer content.
+   * 3. Create the request filename with the file path with respect to the institution code
+   * for the IER 12 Request sent File.
+   * 4. Upload the content to the zoneB SFTP server.
+   * @param generatedDate date in which the assessment for
+   * particular institution is generated.
+   * @returns Processing IER 12 request result.
+   */
+  async processIER12Request(generatedDate?: Date): Promise<IERUploadResult[]> {
+    this.logger.log(
+      `Retrieving pending assessment on ${generatedDate} for IER 12 request...`,
+    );
+    const pendingAssessments =
+      await this.studentAssessmentService.getPendingAssessment(generatedDate);
+    if (!pendingAssessments.length) {
+      return [
+        {
+          generatedFile: "none",
+          uploadedRecords: 0,
+        },
+      ];
+    }
+    this.logger.log(`Found ${pendingAssessments.length} assessments.`);
+    const fileRecords: Record<string, IERRecord[]> = {};
+    pendingAssessments.forEach((pendingAssessment) => {
+      const institutionCode =
+        pendingAssessment.offering.institutionLocation.institutionCode;
+      if (!fileRecords[institutionCode]) {
+        fileRecords[institutionCode] = [];
+      }
+      fileRecords[institutionCode].push(
+        this.createIERRecord(pendingAssessment),
+      );
+    });
+    const uploadResult: IERUploadResult[] = [];
+    try {
+      this.logger.log("Creating IER request content...");
+      for await (const ierUploadResult of Object.keys(fileRecords).map(
```

Comment:
to readability can you add it as a different statement `Object.keys(fileRecords).map(
        (institutionCode) => this.uploadContent(institutionCode, fileRecords)`

Reply:
for await is used in this scenario to restrict the async operation and wait for the result inside the for loop. if we use a map, the uploadResult cannot be populated.